openhcl/page_pool_alloc/src/lib.rs

```diff
@@ -0,0 +1,494 @@
+                base_pfn: range.range.start() / HV_PAGE_SIZE,
+                pfn_bias: addr_bias / HV_PAGE_SIZE,
+                size_pages: range.range.len() / HV_PAGE_SIZE,
+            })
+            .collect();
+
+        Ok(Self {
+            inner: Arc::new(Mutex::new(PagePoolInner {
+                state: pages,
+                device_ids: Vec::new(),
+            })),
+            typ,
+        })
+    }
+
+    /// Create an allocator instance that can be used to allocate pages. The
+    /// specified `device_name` must be unique.
+    ///
+    /// Users should create a new allocator for each device, as the device name
+    /// is used to track allocations in the pool.
+    pub fn allocator(&self, device_name: String) -> anyhow::Result<PagePoolAllocator> {
+        PagePoolAllocator::new(&self.inner, self.typ, device_name)
+    }
+
+    /// Create a spawner that allows creating multiple allocators.
+    pub fn allocator_spawner(&self) -> PagePoolAllocatorSpawner {
+        PagePoolAllocatorSpawner {
+            inner: self.inner.clone(),
+            typ: self.typ,
+        }
+    }
+
+    // TODO: save method and restore
+}
+
+/// A spawner for [`PagePoolAllocator`] instances.
+///
+/// Useful when you need to create multiple allocators, without having ownership
+/// of the actual [`PagePool`].
+#[derive(Debug)]
+pub struct PagePoolAllocatorSpawner {
+    inner: Arc<Mutex<PagePoolInner>>,
+    typ: PoolType,
+}
+
+impl PagePoolAllocatorSpawner {
+    /// Create an allocator instance that can be used to allocate pages. The
+    /// specified `device_name` must be unique.
+    ///
+    /// Users should create a new allocator for each device, as the device name
+    /// is used to track allocations in the pool.
+    pub fn allocator(&self, device_name: String) -> anyhow::Result<PagePoolAllocator> {
+        PagePoolAllocator::new(&self.inner, self.typ, device_name)
+    }
+}
+
+/// A page allocator for memory.
+///
+/// Pages are allocated via the [`Self::alloc`] method and freed by dropping the
+/// associated handle returned.
+#[derive(Debug)]
+pub struct PagePoolAllocator {
+    inner: Arc<Mutex<PagePoolInner>>,
+    #[cfg_attr(not(target_os = "linux"), allow(dead_code))]
+    typ: PoolType,
+    device_id: usize,
+    // TODO: To be used for save/restore. Keep it around just for debuggging,
+    // since otherwise getting the actual name from device_id requires locking
+    // inner.
+    _device_name: String,
+}
+
+impl PagePoolAllocator {
+    fn new(
+        inner: &Arc<Mutex<PagePoolInner>>,
+        typ: PoolType,
+        device_name: String,
+    ) -> anyhow::Result<Self> {
+        let device_id;
+        {
+            let mut inner = inner.lock();
+
+            // device_id must be unique
+            if inner.device_ids.iter().any(|id| id == &device_name) {
+                anyhow::bail!("device name {device_name} already in use");
+            }
+
+            inner.device_ids.push(device_name.clone());
```

Comment:
We never free this.

Reply:
I've actually removed this in the save-restore PR coming next. It's useless as it turns out. I can remove it here if you want, or it's already gone in my tip of tree.

Reply:
No, you're pushing onto the `device_ids` array but you never free.

Reply:
Ah I see. If you drop the allocator, we should remove you from the array. But, the policy is a bit weird - do we allow you to get your previous ID if you create an allocator with the same state? You may have pending allocations that are tracked by a different handle. 

My proposal is this:
1) The device_id is always tied to this device, but if you free the allocator, you can create it again. We'll track Used/Unassigned state internally within the allocator. This is because I do not think with the scheme we have (storing indices inside the allocations themselves) there is any sane/safe way to remove the device_id without walking every allocation. 
2) Outstanding allocations are allowed to stay live, and are still tied to the allocator that may be marked as Unassigned.